### PR TITLE
feat(cli): Support for headless services

### PR DESCRIPTION
**Pull request by gburnard**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #node-monorepo/1770

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Technical improvement or removal of technical debt. 
- [ ] Minor change: comments, documentation, trivial fixes

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [ ] Unit Tests

Optional:

- [ ] Functional Tests
- [ ] Integration Tests

### CODE REVIEW

File: cmd/kubefwd/services/services.go [hunk 1]
```diff
@@ -434,6 +436,60 @@ func getEnv(key, fallback string) string {
...
+	for {
+		event, ok := <-watcher.ResultChan()
+		if !ok {
+			log.Errorf("ENDPOINT: Couldn't receive message")
+			return
+		}
+		log.Warnf("Event: %s", event.Type)
+		switch event.Type {
+		case watch.Added: // add the tunnel
+			log.Warnf("ADDED ENDPOINT: Service %s, Endpoint added. ", service.ObjectMeta.Name)
```
Comment (robblovell) [inline cmd/kubefwd/services/services.go]: I think this should not be a warning or error, just log the statement.

File: pkg/fwdport/fwdport.go [hunk 2]
```diff
@@ -230,9 +230,8 @@ func PortForward(pfo *PortForwardOpts) error {
 	if err = pfo.PortForwardHelper.ForwardPorts(fw); err != nil {
 		log.Errorf("ForwardPorts error: %s", err.Error())
 		pfo.shutdown()
-		//pfo.Stop()
 		return err
-	} else {
+	} else { // service deleted OR kill SIGTERM
 		pfo.shutdown()
 		//pfo.Stop() // Don't shut down, this gives connected clients time to move to a new pod.
```
Comment (robblovell) [inline pkg/fwdport/fwdport.go]: Remove this comment.

File: cmd/kubefwd/services/services.go [hunk 3]
```diff
@@ -470,9 +522,25 @@ func (opts *NamespaceOpts) AddServiceHandler(obj interface{}) {
...
+		ServiceFwd:  svcfwd,
+	}
+
+	if svc.Namespace == "default" && svc.Name == "kubernetes" { // we should not tunnel the k8s API
+		return
+	}
 
-	// Add the service to the catalog of services being forwarded
-	fwdsvcregistry.Add(svcfwd)
+	if selector == "" && svcfwd.Headless {
```
Comment (robblovell) [inline cmd/kubefwd/services/services.go:537]: do we need 'selector == "" &&'?

**Review by robblovell — APPROVED** [review]
Let's talk about the three items here before merging.

File: cmd/kubefwd/services/services.go (hunk 1 above)
Comment (gburnard) [reply cmd/kubefwd/services/services.go]: done

File: pkg/fwdport/fwdport.go (hunk 2 above)
Comment (gburnard) [reply pkg/fwdport/fwdport.go]: done

File: cmd/kubefwd/services/services.go (hunk 3 above)
Comment (gburnard) [reply cmd/kubefwd/services/services.go:537]: added TODO

**Pull request merged**